google/cloud/security/iam/inventory/storage.py

```diff
@@ -559,25 +596,27 @@ def update(self, resource):
         if self.readonly:
             raise Exception('Opened storage readonly')
 
-        Inventory.update_resource(resource)
-        self.buffer.add(resource._row)
-
-    def read(self, key):
-        """Read a resource from the storage.
-
-        Args:
-            key (str): Key of the object to read.
-
-        Returns:
-            object: Row object read from database.
-        """
-
         self.buffer.flush()
-        return (
-            self.session.query(Inventory)
-            .filter(Inventory.index == self.index.id)
-            .filter(Inventory.key == key)
-            .one())
+
+        try:
+            new_rows = Inventory.from_resource(self.index, resource)
+            old_rows = self._get_resource_rows(resource.key())
+
+            new_dict = {row.type_class : row for row in new_rows}
+            old_dict = {row.type_class : row for row in old_rows}
+
+            for type_class in InventoryTypeClass.Supported_TypeClass:
+                if type_class in new_dict:
+                    if type_class in old_dict:
+                        old_dict[type_class].copy_inplace(new_dict[type_class])
+                    else:
+                        old_dict[type_class] = new_dict[type_class]
+                    self.buffer.add(old_dict[type_class])
```

Comment:
buffer performs an add_all, I don't think this will work along with the update.

I think you should specifically use self.session and perform the update here and not use the buffer which is only for adding new objects.